mars/lib/filesystem/_oss_lib/common.py

```diff
@@ -0,0 +1,202 @@
+# Copyright 1999-2021 Alibaba Group Holding Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import base64
+import json
+import os
+import _pyio
+
+from ..base import path_type, stringify_path
+from ....utils import ImportErrorHandler
+
+try:
+	import oss2
+except ImportError:
+	oss2 = ImportErrorHandler('oss2')
+
+# OSS api time out
+_oss_time_out = 10
+
+
+class OSSFileEntry:
+	def __init__(self, path, *, is_dir=None, is_file=None, stat=None, storage_options=None):
+		self._path = path
+		self._name = os.path.basename(path)
+		self._is_file = is_file
+		self._is_dir = is_dir
+		self._stat = stat
+		self._storage_options = storage_options
+	
+	def is_dir(self):
+		if self._is_dir is not None:
+			return self._is_dir
+		self._is_dir = oss_isdir(self._path)
+		return self._is_dir
+	
+	def is_file(self):
+		if self._is_file is not None:
+			return self._is_file
+		if not self.is_dir() or not oss_exists(self._path):
+			self._is_file = False
+		else:
+			self._is_file = True
+		return self._is_file
+	
+	def stat(self):
+		if self._stat is not None:
+			return self._stat
+		self._stat = oss_stat(self._path)
+		return self._stat
+	
+	@property
+	def name(self):
+		return self._name
+	
+	@property
+	def path(self):
+		return self._path
+
+
+def parse_osspath(path: path_type):
+	# Extract OSS configuration from the encoded URL.
+	str_path = stringify_path(path)
+	parse_result = oss2.urlparse(str_path)
+	if parse_result.scheme != "oss":
+		raise ValueError(f"Except scheme oss, but got scheme: {parse_result.scheme}"
+		                 f"in path: {str_path}")
+	bucket = parse_result.hostname
+	if not (parse_result.username and parse_result.password):
+		raise RuntimeError(r"Please use convert_oss_path to add OSS info")
+	param_dict = url_to_dict(parse_result.username)
+	access_key_id = param_dict['access_key_id']
+	access_key_secret = parse_result.password
+	end_point = param_dict['end_point']
+	key = parse_result.path
+	key = key[1:] if key.startswith("/") else key
+	return bucket, key, access_key_id, access_key_secret, end_point
+
+
+def oss_exists(path: path_type):
+	bucket, key, access_key_id, access_key_secret, end_point = parse_osspath(path)
+	oss_bucket = oss2.Bucket(
+		auth=oss2.Auth(access_key_id=access_key_id, access_key_secret=access_key_secret),
+		endpoint=end_point,
+		bucket_name=bucket,
+		connect_timeout=_oss_time_out)
```

Comment:
Shared code fragment across the file. Better to create a common `_get_oss_bucket` utility function.